### PR TITLE
Define an API to query records history

### PR DESCRIPTION
**Pull request by Ndpnt**
- Define and use an API to query records history through `GitRepository` and `MongoRespository`
- Add a migration script as an example of use
- Refactor `storage-adapters` to `repositories` to properly the Repository adn Data Mapper patterns
- Significantly improve performance of git-related queries

### CODE REVIEW

File: scripts/history/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import fsApi from 'fs';
+
+const fs = fsApi.promises;
+
+export async function importReadme({ from: sourceRepository, to: targetRepository }) {
+  const sourceRepositoryReadmePath = `${sourceRepository.path}/README.md`;
+  const targetRepositoryReadmePath = `${targetRepository.path}/README.md`;
+
+  const [readmeCommit] = await sourceRepository.git.log(['README.md']);
```
Comment (MattiSG) [inline scripts/history/utils/index.js:9]: This expects that only a single commit contains the README, which does not sound right to me. I see that you then copy the file contents and get just the message and date of the first commit, which could be understood —if I am right— as squashing all README commits to the first one. This is acceptable behaviour to me but would benefit from being made more explicit.
```suggestion
  const [firstReadmeCommit] = await sourceRepository.git.log(['README.md']);
```

File: scripts/rewrite/initializer/index.js [hunk 2]
```diff
@@ -21,7 +21,10 @@ export async function initReadmeAndLicense(targetRepo, targetPath, authorDate) {
   await fs.copyFile(LICENSE_PATH, targetLicenseFilePath);
   await targetRepo.add(targetReadmeFilePath);
   await targetRepo.add(targetLicenseFilePath);
-  await targetRepo.commit(null, 'Add Readme and License', authorDate);
+  await targetRepo.commit({
+    message: 'Add Readme and License',
```
Comment (MattiSG) [inline scripts/rewrite/initializer/index.js:25]:
```suggestion
    message: 'Add readme and license',
```

File: src/archivist/index.js [hunk 3]
```diff
@@ -41,9 +41,9 @@ export default class Archivist extends events.EventEmitter {
     return Object.keys(this.services);
   }
 
-  constructor({ storage: { versions, snapshots } }) {
+  constructor(config) {
```
Comment (MattiSG) [inline src/archivist/index.js:44]: > Throughout the codebase, this is used as the `config` module, making it here ambiguous.
```suggestion
  constructor(properties) {
```

```suggestion
  constructor(props) {
```

```suggestion
  constructor(options) {
```

File: src/archivist/index.test.js [hunk 4]
```diff
@@ -27,14 +26,12 @@ const VERSIONS_PATH = path.resolve(ROOT_PATH, config.get('recorder.versions.stor
...
-let snapshotsStorageAdapter;
-let versionsStorageAdapter;
-
 async function resetGitRepositories() {
-  return Promise.all([ snapshotsStorageAdapter._removeAllRecords(), versionsStorageAdapter._removeAllRecords() ]);
+  return Promise.all([ app.recorder.snapshotsRepository.removeAll(), app.recorder.versionsRepository.removeAll() ]);
 }
 
 let gitVersion;
+let app;
```
Comment (MattiSG) [inline src/archivist/index.test.js:34]: There is a forward usage of `app` that has not been declared yet. Move these lines right after the `const` declarations.

File: src/archivist/recorder/index.js [hunk 5]
```diff
@@ -1,23 +1,22 @@
-export default class Recorder {
-  constructor({ versionsStorageAdapter, snapshotsStorageAdapter }) {
-    if (!versionsStorageAdapter || !snapshotsStorageAdapter) {
-      throw new RangeError('Storage adapters should be defined both for versions and snapshots');
```
Comment (MattiSG) [inline src/archivist/recorder/index.js:4]: > This validation does not seem to have been reproduced in this class nor in the Factory. It was not tested. I'm okay with losing it and considering behaviour undefined in case storage is not defined for both cases, but this is less defensive as it was.

File: src/archivist/recorder/record.test.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import chai from 'chai';
+import config from 'config';
+
+import Record from './record.js';
+import RepositoryFactory from './repositories/factory.js';
+
+const { expect } = chai;
+
+describe('Record', () => {
```
Comment (MattiSG) [inline src/archivist/recorder/record.test.js:9]: We are missing a constructor / validator test.

File: src/archivist/recorder/repositories/git/dataMapper.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+export const COMMIT_MESSAGE_PREFIXES_REGEXP = new RegExp(`^(${COMMIT_MESSAGE_PREFIX.startTracking}|${COMMIT_MESSAGE_PREFIX.refilter}|${COMMIT_MESSAGE_PREFIX.update})`);
+export default class GitDataMapper {
+  static toPersistence(record, prefixMessageToSnapshotId) {
+    const { serviceId, documentType, isRefilter, snapshotId, mimeType, isFirstRecord } = record;
+
+    let prefix = isRefilter ? COMMIT_MESSAGE_PREFIX.refilter : COMMIT_MESSAGE_PREFIX.update;
+
+    prefix = isFirstRecord ? COMMIT_MESSAGE_PREFIX.startTracking : prefix;
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/git/dataMapper.js:22]:
```suggestion
    let prefix = isFirstRecord ? COMMIT_MESSAGE_PREFIX.startTracking :
    isRefilter ? COMMIT_MESSAGE_PREFIX.refilter :
    COMMIT_MESSAGE_PREFIX.update;
```

File: src/archivist/recorder/repositories/git/dataMapper.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+mime.define({ 'text/markdown': ['md'] }, true); // ensure extension for markdown files is `.md` and not `.markdown`
+
+export const COMMIT_MESSAGE_PREFIX = {
+  startTracking: 'Start tracking',
+  refilter: 'Refilter',
+  update: 'Update',
+};
+
+export const COMMIT_MESSAGE_PREFIXES_REGEXP = new RegExp(`^(${COMMIT_MESSAGE_PREFIX.startTracking}|${COMMIT_MESSAGE_PREFIX.refilter}|${COMMIT_MESSAGE_PREFIX.update})`);
+export default class GitDataMapper {
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/git/dataMapper.js:16]: There are only static members: why is this a class?

File: src/archivist/recorder/repositories/git/git.js [hunk 9]
```diff
@@ -114,6 +104,18 @@ export default class Git {
     return this.git.raw(options);
   }
 
+  async show(options) {
+    return this.git.show(options);
+  }
+
+  async getFullHash(shortHash) {
+    return (await this.git.show([ shortHash, '--pretty=%H', '-s' ])).trim();
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/git/git.js:112]: I don't understand how that would work in cases where the short hash is ambiguous 🤔

File: src/archivist/recorder/repositories/git/index.js [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+import RepositoryInterface from '../interface.js';
+
+import DataMapper, { COMMIT_MESSAGE_PREFIXES_REGEXP, COMMIT_MESSAGE_PREFIX } from './dataMapper.js';
+import Git from './git.js';
+
+const fs = fsApi.promises;
+const PDF_MIME_TYPE = 'application/pdf';
+
+mime.define({ 'text/markdown': ['md'] }, true); // ensure extension for markdown files is `.md` and not `.markdown`
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/git/index.js:19]: This is repeated in https://github.com/ambanum/OpenTermsArchive/pull/875/files#diff-6673716d44e6e04b6852f1537f050fa7ebf0bf5c2c7f55d27a6f2de7f627ddadR7.

File: src/archivist/recorder/repositories/git/index.js [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+    return record;
+  }
+
+  finalize() {
+    if (!this.needsPublication) {
+      return;
+    }
+
+    return this.git.pushChanges();
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/git/index.js:61]: That method is `async`, how can this one not be?

File: scripts/dataset/export/index.js [hunk 12]
```diff
@@ -16,7 +17,7 @@ const fs = fsApi.promises;
 const ARCHIVE_FORMAT = 'zip'; // for supported formats, see https://www.archiverjs.com/docs/archive-formats
 
 export default async function generate({ archivePath, releaseDate }) {
-  const versionsStorageAdapter = await (instantiateVersionsStorageAdapter()).initialize();
+  const versionsRepository = await (RepositoryFactory.create(config.get('recorder.versions.storage'))).initialize();
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply scripts/dataset/export/index.js:20]: It's an error.

File: src/archivist/recorder/record.test.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+  let repository;
+  let record;
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/record.test.js:11]: I will change it here for consistency, but we should have a discussion about this, I am not as convinced as before.

File: src/archivist/recorder/repositories/git/dataMapper.js (hunk 7 above)
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/dataMapper.js:22]: Nested ternary expression can be hard to read so we choose to disallow them with https://eslint.org/docs/rules/no-nested-ternary

File: src/archivist/recorder/repositories/git/git.js [hunk 14]
```diff
@@ -32,35 +32,44 @@ export default class Git {
...
-  async commit(filepath, message, authorDate) {
-    const options = {};
+  async commit({ filepath, message, date = new Date() }) {
+    let summary;
 
-    if (authorDate) {
-      options['--date'] = new Date(authorDate).toISOString();
-    }
+    try {
+      const commitDate = new Date(date).toISOString();
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/git.js:39]: I not really agree. As the finally is here to rollback env setup even in case of error, I prefer to have the `try` wrapping the content related to the `finally`.

File: src/archivist/recorder/repositories/git/git.js (hunk 9 above)
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/git.js:112]: It won't… and it can't. That's why I wanted to switch to full hash.

**Review by MattiSG — COMMENTED** [review]
Almost done.

File: src/archivist/recorder/repositories/git/index.js (hunk 10 above)
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/index.js:19]: Yep I knew, currently both are needed. So what do you suggest?

File: src/archivist/recorder/repositories/git/index.js (hunk 11 above)
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/index.js:61]: When marking method `async` it will just wrap it in a promise. In this case I return a promise, so there is no need to wrap it.

File: src/archivist/recorder/repositories/mongo/index.js [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+import { MongoClient, ObjectId, Binary } from 'mongodb';
+
+import RepositoryInterface from '../interface.js';
+
+import DataMapper from './dataMapper.js';
+
+export default class MongoRepository extends RepositoryInterface {
+  constructor({ database: databaseName, collection: collectionName, connectionURI }) {
+    super();
+    const client = new MongoClient(connectionURI);
```
Comment (MattiSG) [inline src/archivist/recorder/repositories/mongo/index.js:15]: Shouldn't the client be a private variable?
```suggestion
    this.#client = new MongoClient(connectionURI);
```

**Review by MattiSG — COMMENTED** [review]
First pass done.

File: src/archivist/recorder/repositories/git/git.js (hunk 9 above)
Comment (MattiSG) [reply src/archivist/recorder/repositories/git/git.js:112]: I get that, but the way we get the full hash is by getting the source commit through its short hash, so won't a collision always prevent `getFullHash` from working?

File: scripts/dataset/export/index.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (MattiSG) [reply scripts/dataset/export/index.js]: > What do you suggest instead?

Shouldn't `create` always call `initialise`? 🤔

**Review by MattiSG — DISMISSED** [review]
👍

File: scripts/history/migrate-services.js [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+  console.timeEnd('Total time');
+
+  await finalize(migration);
+}());
+
+async function rewriteSnapshots(repository, records, idsMapping, logger) {
+  let i = 1;
+
+  for (const record of records) {
+    const { id: recordId } = await repository.record(record); // eslint-disable-line no-await-in-loop
```
Comment (Ndpnt) [inline scripts/history/migrate-services.js:143]: Use .save instead of .record

File: scripts/history/logger/index.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+      if (type.length > 50) {
+        type = `${type.substring(0, 48)}…`;
+      }
+
+      prefix += `${type}`.padEnd(50);
+    }
+
+    if (id) {
+      prefix += `${id}`.padEnd(42);
+    }
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply scripts/history/logger/index.js:31]: Yes but it's not very important. This logger is only used in history scripts and either we have all the parameters or we only have the message.

File: scripts/history/migrate-services.js [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+    prefixMessageToSnapshotId: 'This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/france-snapshots/commit/',
+  },
+  to: {
+    snapshots: 'france-elections-snapshots',
+    versions: 'france-elections-versions-hash-updated',
+    prefixMessageToSnapshotId: 'This version was recorded after filtering snapshot https://github.com/OpenTermsArchive/france-elections-snapshots/commit/',
+  },
+};
+
+const COUNTERS = {
```
[the remark this answers is not in the document]
Comment (MattiSG) [reply scripts/history/migrate-services.js:31]: Yes, I agree with that understanding. I'm not saying it's a problem, I just find it weird: the hint given to my mental model, both with `const` and all caps, is that the intent of this variable is to represent a constant, and I find it hard to think of a constant as just a reference and then all its members are intended to be mutable.

[the remark this answers is not in the document]
Comment (Ndpnt) [reply scripts/history/migrate-services.js:31]: I understand and I agree. I think we should keep all caps to effectively constant values, I will rename `COUNTERS` to `counters`.

File: scripts/history/migrate-services.js [hunk 19]
```diff
@@ -0,0 +1,212 @@
+import fsApi from 'fs';
```
Comment (martinratinaud) [inline scripts/history/migrate-services.js:1]: This implementation seems very clear to me but have the downside that every migration created will have commitable changes.

I think this may not be part of this PR but it would be great to have the config loaded from a gitignored file `.migration.json`
Only change needed would be to add the type of repository to it so that the right repository class is used

File: src/archivist/recorder/repositories/git/git.js (hunk 9 above)
Comment (MattiSG) [reply src/archivist/recorder/repositories/git/git.js:112]: We agreed on a synchronous call to have a better implementation in a separate PR.

File: src/archivist/recorder/index.test.js [hunk 20]
```diff
@@ -51,20 +23,22 @@ describe('Recorder', () => {
         let record;
 
         before(async () => {
-          recorder = new Recorder({
-            versionsStorageAdapter: versionsAdapter,
-            snapshotsStorageAdapter: snapshotsAdapter,
-          });
+          const options = config.util.cloneDeep(config.recorder);
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/index.test.js:26]: Because the object returned by `node-config` is immutable and I need to modify it.

File: src/archivist/recorder/repositories/git/git.js [hunk 21]
```diff
@@ -32,35 +32,44 @@ export default class Git {
...
 
-    let summary;
+      process.env.GIT_AUTHOR_DATE = commitDate;
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/repositories/git/git.js:41]: Because it is the only way to set the git committer date.

File: src/archivist/recorder/repositories/interface.js [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+  * @param {string} recordId - Record ID of the record to find
+  * @returns {Promise<Record>} Promise that will be resolved with the found record or an empty object if none match the given ID
+  */
+  async findById(recordId) {
+    throw new Error(`#findById method is not yet implemented in ${this.constructor.name}`);
+  }
+
+  /**
+  * Find all records
+  * For performance reasons, the content of the records will not be loaded by default. Use #loadRecordContent to load the content of individual records
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/repositories/interface.js:60]: I think the previous wording is clearer as the caller will have to load the content of each record found

File: src/archivist/recorder/repositories/mongo/index.js [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+      return record;
+    }
+
+    await this.loadRecordContent(record);
+
+    return record;
+  }
+
+  async #toPersistence(record) {
+    if (record.content === undefined || record.content === null) {
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/repositories/mongo/index.js:117]: Content could be defined manually so it could be undefined or null

File: src/archivist/index.js [hunk 24]
```diff
@@ -41,9 +41,9 @@ export default class Archivist extends events.EventEmitter {
     return Object.keys(this.services);
   }
 
-  constructor(config) {
+  constructor({ recorder }) {
```
Comment (MattiSG) [inline src/archivist/index.js:44]: > I find this less clear than knowing the parameter is called `config`. I expect an instance of `Recorder` in the `recorder` field, not a config object for a new Recorder.

File: src/archivist/recorder/index.test.js [hunk 25]
```diff
@@ -32,10 +32,7 @@ describe('Recorder', () => {
           await recorder.initialize();
         });
 
-        after(async () => {
-          await recorder.snapshotsRepository.removeAll();
-          await recorder.finalize();
-        });
+        after(async () => recorder.finalize());
```
Comment (MattiSG) [inline src/archivist/recorder/index.test.js:35]:
```suggestion
        after(recorder.finalize.bind(recorder));
```

**Review by MattiSG — COMMENTED** [review]
I'm impressed that we can drop `src/index.js` altogether. With these incremental reviews, I can't really be sure how appropriate it is.

File: src/archivist/recorder/repositories/mongo/dataMapper.js [hunk 26]
```diff
@@ -2,45 +2,30 @@ import { ObjectId } from 'mongodb';
...
-    documentFields.created_at = new Date();
-
-    return documentFields;
+  if (documentFields.snapshotId) {
+    documentFields.snapshotId = new ObjectId(record.snapshotId);
   }
 
-  static toDomain(document) {
-    const { _id, serviceId, documentType, fetchDate, mimeType, isRefilter, isFirstRecord, snapshotId } = document;
+  documentFields.content = record.content;
```
[the remark this answers is not in the document]
Comment (Ndpnt) [reply src/archivist/recorder/repositories/mongo/dataMapper.js:12]: No because `content` is defined with a getter and it's not enumerable, so it is not defined in `documentFields`
